app/src/main/java/org/dicekeys/app/openpgp/Mpi.kt

```diff
@@ -0,0 +1,41 @@
+package org.dicekeys.app.openpgp
+
+import com.google.common.io.ByteStreams
+import java.math.BigInteger
+
+//  Unencrypted Multiprecision Integers
+class Mpi(val value: BigInteger) {
+    val size : UShort by lazy {
+        value.bitLength().toUShort()
+    }
+
+    fun toByteArray(): ByteArray{
+        val body = ByteStreams.newDataOutput()
+        body.writeShort(size.toInt())
+        if(size > 0u) {
+            body.write(value.toByteArray().let {
+                // BigInteger representation always have a sign bit. This is not a problem as we only
+                // handle positive numbers. The only edge case is when we use eg. all 8 bits of a byte
+                // to represent a positive number. BigInteger will have to store (8 number bits + 1 sign bits)
+                // 1 byte fits the number representation and it will need to consume one more zero byte in our case
+                // to represent the sign positive bit.
+                // In that case it's safe to remove the first byte if it's 0
+                if(it[0] == 0.toByte()){
+                    return@let it.takeLast(it.size - 1).toByteArray()
+                }
+                it
+            })
+        }
+        return body.toByteArray()
+    }
+
+    companion object{
+        fun fromHex(hex: String): Mpi {
+            return Mpi(BigInteger(hex, 16))
+        }
+
+        fun fromByteArray(byteArray: ByteArray): Mpi {
+            return Mpi(BigInteger(1, byteArray))
```

Comment:
In doing c++ conversion, I was trying to understand what this [BigInteger constructor does](https://developer.android.com/reference/java/math/BigInteger).

This looks like it will force whatever number is in byteArray to be treated as a negative number.  Why is this used and not just BigInteger(byteArray)?  Won't this cause positive numbers to become negative numbers?

Reply:
The spec for [MPI](https://datatracker.ietf.org/doc/html/rfc4880#section-3.2)

`1` is for positive number. I am doing that cause the constructor without signum, takes the first byte of the byte array as the sign of the number. Take a look into the note in `toByteArray()` where it handles a case related to BigNumber and sign.

If in C++ could store a similar value without the sign it's a better fit. Unfortunately BigInteger always has a sign bit.

Reply:
So i am forcing BigInteger to handle the value as a positive number, so now we have an extra sign bit (zero for positive number) and is important to handle that extra sign bit when converting it back to byte array as i am doing in the `toByteArray()`.

Reply:
For your test case:
"58CBA8496EABC3D58F84C034448EF1C1F95C9C6582E006C2BB205B70EB58D5CF"
You have  
"00FF58CBA8496EABC3D58F84C034448EF1C1F95C9C6582E006C2BB205B70EB58D5CF"

For a new test case, where we remove the "BA" and assert "00" up front...
"0058A8496EABC3D58F84C034448EF1C1F95C9C6582E006C2BB205B70EB58D5CF"
Our length word would presumably be "00F7" (8 bits smaller) since the 00 is up front.
Would we then skip the first byte and output, as you do in the test case for 0, and have a smaller packet size?
"00F758CBA8496EABC3D58F84C034448EF1C1F95C9C6582E006C2BB205B70EB58D5CF"

I'd suggest there should be a test case for this.